### ISSUE TITLE
[RFC] Improve performance for softmax op for cuda in some specific size

### ISSUE THREAD

**Issue by ywq880611** [pytorch/pytorch#144645]
### 🚀 The feature, motivation and pitch

In the triton [tutorial](https://triton-lang.org/main/getting-started/tutorials/02-fused-softmax.html) for softmax, we could compare the performance of pytorch op and triton, here is the result on my local (RTX3080):
### Compare between triton and torch
![softmax-perf](https://github.com/user-attachments/assets/eb24461e-66a2-4738-bcdd-348f052ba993)
We could see there is a dramatic perf drop after **N=1024** for torch.
### Mini repro
Here is my mini repro to test the perf around **N=1024**
```python
import torch

DEVICE=torch.device('cuda')

# Time cost for near 1024
for cnt in range(1020, 1030):
    x = torch.randn(4096, cnt, device=DEVICE, dtype=torch.float32)
    #x = torch.randn(M, N, device=DEVICE, dtype=torch.float32)

    #warm up
    need_warmup = True
    round = 5
    if need_warmup:
        for _ in range(round):
            output = torch.softmax(x, dim=-1)

    torch.cuda.synchronize()
    start_time = torch.cuda.Event(enable_timing=True)
    end_time = torch.cuda.Event(enable_timing=True)

    # Start time
    start_time.record()

    # Apply softmax
    for _ in range(round):
        output = torch.softmax(x, dim=-1)

    # End time
    end_time.record()

    torch.cuda.synchronize()

    # Calculate elapsed time
    elapsed_time_ms = start_time.elapsed_time(end_time)
    # print(f"CUDA Time: {elapsed_time_ms:.6f} ms")
    gbps = lambda ms: round * 2 * x.numel() * x.element_size() * 1e-9 / (ms * 1e-3)
    print(f"n as {cnt} of softmax: {gbps(elapsed_time_ms):.6f} gb/s")
```

Its output is:
```
n as 1020 of softmax: 645.059274 gb/s
n as 1021 of softmax: 653.439969 gb/s
n as 1022 of softmax: 644.096473 gb/s
n as 1023 of softmax: 649.523815 gb/s
n as 1024 of softmax: 656.015990 gb/s
n as 1025 of softmax: 209.183680 gb/s
n as 1026 of softmax: 208.490244 gb/s
n as 1027 of softmax: 201.126073 gb/s
n as 1028 of softmax: 278.307944 gb/s
n as 1029 of softmax: 205.510996 gb/s
```
We could see there is about ***>3x*** perf drop after **n=1024** `(656 vs 209)`.

#### Current state
Let's look at the below code snippet:
https://github.com/pytorch/pytorch/blob/1664033e13cc8831e2bb66e5c975ffb4dfc24eda/aten/src/ATen/native/cuda/SoftMax.cu#L846-L880
There are two kinds of softmax op in CUDA:
1. `dispatch_softmax_forward`
2. `cunn_SoftMaxForwardSmem` or `cunn_SoftMaxForward`

The implementation of `cunn_SoftMaxForwardSmem` or `cunn_SoftMaxForward` will be invoked if **N>1024**, but it's not efficient in the rough range 1025~2000. 

#### root cause
The reason why the `cunn_SoftMaxForwardSmem` and `cunn_SoftMaxForward` is not efficient is **too much global memory and share memory access**, because they didn't cache the data each thread would like to access in register, so they may have to load data from memory much times.

### Alternatives

We could try to use registers to cache the data which a thread would like to use.

Pros: 
Improve performance by reducing memory access
Cons: 
Increase register pressure, but we could just do it in range of `N` in about 1025~2000, which may just increase about `1~2` register for a thread (I guess it's acceptable).

Now I have a draft implement in my local, it shows **~50%** performance gain compared with current torch, for the above case, its output is:
```
n as 1020 of softmax: 613.605899 gb/s
n as 1021 of softmax: 628.307691 gb/s
n as 1022 of softmax: 627.565386 gb/s
n as 1023 of softmax: 658.589191 gb/s
n as 1024 of softmax: 632.968698 gb/s
n as 1025 of softmax: 297.640651 gb/s
n as 1026 of softmax: 297.408139 gb/s
n as 1027 of softmax: 298.221413 gb/s
n as 1028 of softmax: 279.757637 gb/s
n as 1029 of softmax: 298.785226 gb/s
```
I'm still working on to make it better now.

WDYT? Any insights or comments are appreciated!

### Additional context

A [doc](https://docs.google.com/document/d/1K030-wgNlzyYePBAPvwZ5W0Tm7jebFAjSat6QvmWoz4/edit?usp=sharing) contains some nsight profiler screen shot.

cc @msaroufim @ptrblck @eqy

**Comment by jbschlosser** [comment]
Triage review for labeling + visibility

**Comment by eqy** [comment]
What performance do you observe when changing the heuristic `if (dim_size <= 1024 && dim_size*sizeof(scalar_t) <= 4096) {` to gate on a value larger than 1024? Is there a problem with the amount of shared memory requested?

**Comment by ywq880611** [comment]
Hi @eqy ,

Thank you for your comments! IIUC, we couldn't directly tune the heuristic, to make `dispatch_softmax_forward` works when `dim_size > 1024`, because this kernel will utilize a warp to reduce a vector, each thread in this warp could just process **32 elements at most**, so this kernel could just process `32 * 32 = 1024` elements at most. 
The code comment seems also mention that:
https://github.com/pytorch/pytorch/blob/1664033e13cc8831e2bb66e5c975ffb4dfc24eda/aten/src/ATen/native/cuda/PersistentSoftmax.cuh#L46-L47

There is another alternative in my mind is to expand this kernel, to make it work properly when `N > 1024`, like N in range (1025, 2048), because the performance was suffering in this range, but it may increase more register pressure, it may need 32 additional registers at most. WDYT?

**Comment by ywq880611** [comment]
Update:

I also expand the `dispatch_softmax_forward` kernel in this [commit](https://github.com/pytorch/pytorch/pull/144679/commits/0c6a2494386cf53a378b73af98e0086480b15324) in range(1024, 2048), now the perf landscape compared with `triton` looks like:

![Image](https://github.com/user-attachments/assets/45a56d47-d756-4041-a1fc-029818a86a38)
We could see there is perf improvement two ranges:
1. Range(1024, 2048):
We use the `dispatch_softmax_forward` for this range, it could achive on par perf compared with `triton`, so we got **>3x** perf gain in this range. but it may introduce more register pressure (for `N = 1025`, it uses **96 registers**, for `N = 1024`, it uses **56 registers**.)
2. Range(2048, ~):
There is also performance gain with new kernel `cunn_SoftMaxForwardReg`, it also bring benefits to this op, but as the size of N increase, the benefits will be decreased, it could get **~50%** perf gain when near 2048:
baseline:
    ```
    n as 2045 of softmax: 332.520311 gb/s
    n as 2046 of softmax: 330.666676 gb/s
    n as 2047 of softmax: 329.403774 gb/s
    n as 2048 of softmax: 499.726461 gb/s
    n as 2049 of softmax: 322.995059 gb/s
    n as 2050 of softmax: 321.253681 gb/s
    n as 2051 of softmax: 323.509547 gb/s
    n as 2052 of softmax: 484.426425 gb/s
    n as 2053 of softmax: 321.026178 gb/s
    n as 2054 of softmax: 317.968202 gb/s
    ```
    `cunn_SoftMaxForwardReg` kernel:
    ```
    n as 2045 of softmax: 618.525506 gb/s
    n as 2046 of softmax: 650.815124 gb/s
    n as 2047 of softmax: 648.554451 gb/s
    n as 2048 of softmax: 651.613203 gb/s
    n as 2049 of softmax: 484.410597 gb/s
    n as 2050 of softmax: 487.234248 gb/s
    n as 2051 of softmax: 483.588297 gb/s
    n as 2052 of softmax: 492.741751 gb/s
    n as 2053 of softmax: 495.655227 gb/s
    n as 2054 of softmax: 486.333687 gb/s
    ```
